modules/ldap/src/main/java/com/neighborhoods/wasabi/ldap/impl/LdapUserDirectory.java

```diff
@@ -0,0 +1,240 @@
+     * @param <V> the value type
+     * @param map the map
+     * @return the stream
+     */
+    private static <K, V> Stream<V> mapToStream(Map<K, V> map) {
+        return map.values().stream();
+    }
+
+    /** The cached users. */
+    protected ConcurrentHashMap<String, DirectoryUser> users;
+
+    /** The directory delegate. */
+    protected final DirectoryDelegate delegate;
+
+    /**
+     * Instantiates a new ldap user directory.
+     *
+     * @param delegate the delegate
+     */
+    @Inject
+    public LdapUserDirectory(DirectoryDelegate delegate) {
+        this.delegate = delegate;
+        refreshCache();
+    }
+
+    /**
+     * Add user to local cache
+     * 
+     * @param user The directory user to add to the cache
+     * @return The result of the cache put operation
+     */
+    @Override
+    public DirectoryUser addUserToCache(DirectoryUser user) {
+        return this.users.put(user.getUsername().getUsername(), user);
+    }
+
+    /**
+     * Authenticate (e.g. during initial login) a given user
+     * 
+     * @param username - The username to authenticate
+     * @param password - The password to authenticate
+     * @return The directory user if successful
+     */
+    @Override
+    public DirectoryUser authenticate(String username, String password) {
+        return delegate.authenticate(this, username, password);
+    }
+
+    /**
+     * Get a stream of all users
+     * 
+     * @return All directory users
+     */
+    @Override
+    public Stream<DirectoryUser> getAllUsers() {
+        Stream<DirectoryUser> result = null;
+        result = mapToStream(this.users);
+        return result;
+    }
+
+    /**
+     * Gets the directory user from the cache.
+     *
+     * @param username The target user's username
+     * @return The directory user object from cache
+     */
+    protected DirectoryUser getDirectoryUserFromCache(String username) {
+        DirectoryUser cachedUser;
+        cachedUser = this.users.get(username);
+
+        if (cachedUser != null) {
+            return validateCachedUser(cachedUser);
+        }
+        return null;
+    }
+
+    /**
+     * Validate a directory token
+     * 
+     * @param username The username to validate
+     * @param encryptedPassword The password to validate
+     * @return true, if a directory token is valid
+     */
+    @Override
+    public boolean isDirectoryTokenValid(String username, String encryptedPassword) {
+        return delegate.isDirectoryTokenValid(this, username, encryptedPassword);
+    }
+
+    /**
+     * Find a user by username
+     * 
+     * @param username The target user's username
+     * @return The user object
+     */
+    @Override
+    public DirectoryUser lookupDirectoryUser(String username) {
+        DirectoryUser result = getDirectoryUserFromCache(username);
+        if (result != null) {
+            return result;
+        }
+        return delegate.getDirectoryUserByUsername(this, username);
```

Comment:
so in the case where we don't have it in the cache but do get a real user back from the delegate, do we not want to also re-add that to the cache?
same question for `lookupUser` and `lookupUserByEmail` below

Reply:
ah ok i get it.
cool. that's good stuff.